misc/package/Data Files/MWSE/core/mcm/fileUtils.lua

```diff
@@ -0,0 +1,71 @@
+-- Fetches Components and Variables by class name.
+-- Intended for internal use only.
+local fileUtils = {}
+
+
+
+local prefixLength = string.len("Data Files\\MWSE\\core\\")
+
+-- Store component/variable paths so we only have to iterate the directory once. 
+-- (Testing has shown this results in a noticeable boost in performance.)
+
+local componentPaths = {}
+local variablePaths = {}
+
+for filePath, dir, fileName in lfs.walkdir("data files\\mwse\\core\\mcm\\components\\") do
+	-- For example, when adding the path of `mwseMCMPage`:
+	-- The key   will be:  "Page" 
+	--        instead of:  "Page.lua"
+	-- The value will be:  "mcm.components.pages.Page" 
+	-- 		  instead of:  "data files\\mwse\\core\\mcm\\components\\pages\\Page.lua"
+	componentPaths[fileName:sub(1, fileName:len() - 4)] = filePath:sub(1 + prefixLength, filePath:len() - 4):gsub("[\\/]", ".")
+end
+
+for filePath, dir, fileName in lfs.walkdir("data files\\mwse\\core\\mcm\\variables\\") do
+	variablePaths[fileName:sub(1, fileName:len() - 4)] = filePath:sub(1 + prefixLength, filePath:len() - 4):gsub("[\\/]", ".")
+end
+
+local components = {}
+local variables = {}
+
+---@protected
+---@param className string The name of the class of the `mwseMCMComponent` to search for.
+---@return mwseMCMComponent?
+function fileUtils.getComponentClass(className)
+	local class = components[className]
+	if class then return class end
+	if className == "HyperLink" then
```

Comment:
Some mods wanted to call `createHyperLink` instead of `createHyperlink`, so I hardcoded a safeguard into this function. Same thing happened with `createSidebarPage` instead of `createSideBarPage`.

A more general solution might be preferable though.

Reply:
> Maybe it's better to store things in the components and variables tables by first calling string.lower on the keys? And then calling string.lower on the className parameter

You've just surpassed me 😁.

About the performance impact, I can't really say. As always, performance should be measured before reasoning. Nevertheless, I'd take this in even if it doesn't end up faster than the master because it reduces code duplication.

Reply:
The performance advice is very sound.

I thought of another solution that doesn't use `string.lower`, although it still feels like it may be a little unstable: I can remove the cursed `if className == ...` checks by adding the following two lines:
```lua
componentPaths.HyperLink = componentPaths.Hyperlink
componentPaths.SidebarPage = componentPaths.SideBarPage
```

i.e., specifically mention legacy names for certain components at the `componentPaths` level.

Here are some pros and cons to this approach (compared to the `string.lower` one):

Not using `string.lower` means we don't support weird capitalization that isn't officially documented, i.e. `createTEMPLaTE` or something. Each of the weird capitalizations would result in an extra function being generated by the `__index` metamethods and added to `Category`, `Template` and/or `mcm`. So it may be nice to limit for that reason. (Although I really doubt people would even think to try something like `mwse.mcm.createTEMPLaTE` unless they specifically knew they could. But it might be a bit weird that the `"create"` part needs to be lowercaes, while the `className` part doesn't.)

The downside is that this approach would theoretically be less stable than using `string.lower`, because it really depends on the capitalization of the file names. And this part makes me nervous. But I'm also not sure how much of a difference this would make, since the file names should (theoretically) always be capitalized in the same way.

Reply:
I think we are fine as is.